civismlext/hyperband.py

```diff
@@ -298,45 +299,78 @@ def fit(self, X, y=None, groups=None, **fit_params):
         out = []
         smax = int(np.floor(np.log(R / Rmin) / np.log(self.eta)))
         B = (smax + 1.0) * R
+
+        # This code is hyperband, but I have swapped the order of the
+        # inner and outer loops to expose more parallelism. Fun.
+        Ts = []
+        ns = []
+        rs = []
         for s in range(smax, -1, -1):
-            n = int(np.ceil(B / R * np.power(self.eta, s) / (s + 1.0)))
-            r = int(R / np.power(self.eta, s))
-            T = list(ParameterSampler(self.param_distributions,
-                                      n,
-                                      random_state=self._random_state))
-
-            for i in range(0, s + 1):
-                n_i = int(np.floor(n / np.power(self.eta, i)))
-                r_i = int(r * np.power(self.eta, i))
-
-                _jobs = []
-                for parameters in T:
-                    _parameters = copy.deepcopy(parameters)
-                    _parameters.update(
-                        {list(self.cost_parameter_max.keys())[0]: r_i})
-                    for train, test in cv_iter:
-                        _jobs.append(delayed(_fit_and_score)(
-                            clone(base_estimator), X, y, self.scorer_,
-                            train, test, self.verbose, _parameters,
-                            fit_params=fit_params,
-                            return_train_score=self.return_train_score,
-                            return_n_test_samples=True,
-                            return_times=True, return_parameters=True,
-                            error_score=self.error_score))
-
-                _out = Parallel(
-                    n_jobs=self.n_jobs, verbose=self.verbose,
-                    pre_dispatch=pre_dispatch)(_jobs)
-
-                results, _ = self._process_outputs(_out, n_splits)
+            ns.append(int(np.ceil(B / R * np.power(self.eta, s) / (s + 1.0))))
+            rs.append(int(R / np.power(self.eta, s)))
+            Ts.append(list(ParameterSampler(
+                self.param_distributions,
+                ns[-1],
+                random_state=self._random_state)))
+        nums = copy.copy(ns)
+        offsets = [0] + list(
+            np.cumsum(np.array(nums) * n_splits).astype(int))[:-1]
+
+        for rnd in range(0, smax + 1):
```

Comment:
I think we should clarify why we're iterating the max times over everything. Maybe a note here like "iterate the maximum number of times for each resource budget configuration. If we should skip an interation, T will be an empty list".